### PR TITLE
Use 'viewId' as progress location

### PR DESCRIPTION
**Pull request by RomanNikitenko**
#### What it does

At the moment it's possible to use [ProgressLocation](https://github.com/eclipse-theia/theia/blob/bdc9aa276cb55e5cf3dc8d7cb6e795ed8ffba998/packages/plugin/src/theia.d.ts#L4172-L4185) enum only as a location at which progress should show.
The PR is aligning [location field](https://github.com/eclipse-theia/theia/blob/bdc9aa276cb55e5cf3dc8d7cb6e795ed8ffba998/packages/plugin/src/theia.d.ts#L4154) to [the corresponding field](https://github.com/microsoft/vscode/blob/ad1dc7f87a94edc1d06cbeb2261459f2764b2c43/src/vs/vscode.d.ts#L9181) of VS Code to provide an opportunity to use `viewId` as progress location.

#### How to test

I created a [test plugin](https://github.com/RomanNikitenko/progress-test-plugin), it contains [commands](https://github.com/RomanNikitenko/progress-test-plugin/blob/bbed24e827e73ee4ae41391010eeb3abb7ea0f25/src/progress-test-plugin-backend.ts#L31-L57) for testing
1. Please clone the plugin and build it using yarn.
2. Copy the plugin to `theia/plugins` folder. 
3. Use the following commands from Command Palette (F1) for testing: 
- Start progress for Explorer view
- Start progress for Search view
- Start progress for Debug view
- Start progress for SCM view

First 3 commands use `viewId` (new behavior) to show progress while running the given callback, the last command uses [ProgressLocation](https://github.com/eclipse-theia/theia/blob/bdc9aa276cb55e5cf3dc8d7cb6e795ed8ffba998/packages/plugin/src/theia.d.ts#L4172-L4185) enum to show progress for SCM view (existed behavior). 

progress_API

4. Use case for a custom view is described here https://github.com/eclipse-theia/theia/pull/8700#issuecomment-721078069 - thanks a lot @kittaakos for the case and the extension for testing 
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Roman Nikitenko <rnikiten@redhat.com>

### CODE REVIEW

**Review by benoitf — APPROVED** [review]
Tested 4 commands with the provided plug-ins
I see progress as expected on 4 examples

image

More compliance with VS Code 🎉

**Comment by kittaakos** [comment]
I also give it a try...

File: packages/plugin-ext/src/plugin/notification.ts
Comment (kittaakos) [inline packages/plugin-ext/src/plugin/notification.ts:52]: VS Code [does not do a check](https://github.com/microsoft/vscode/blob/7d1affcf90de1559b6d6f8da01eac75d3b735750/src/vs/workbench/api/common/extHostTypeConverters.ts#L1145-L1147) for `location.viewId`.
What should happen when I pass in the following argument: `{ viewId: '' }`? It will evaluate to `falsy` and we hit the `switch`. Can you please remove the extra `location.viewId` check and align the code with Code 😊 ?

**Comment by kittaakos** [comment]
> I also give it a try...

This does not seem to work with a custom view inside a custom view container. I will try to come up with a public, stripped-down example later today. Is this urgent to merge, @RomanNikitenko?

File: packages/plugin-ext/src/plugin/notification.ts
Comment (RomanNikitenko) [reply packages/plugin-ext/src/plugin/notification.ts]: thanks, I removed additional check for `location.viewId`

**Comment by RomanNikitenko** [comment]
> This does not seem to work with a custom view inside a custom view container. I will try to come up with a public, stripped-down example later today. Is this urgent to merge, @RomanNikitenko?

I tested the solution for `explorer`, `debug`, `search` and `scm` views only, so I didn't face the problem related to a custom view inside a custom view container. Please provide steps to reproduce or an example for testing.

About 
> Is this urgent to merge

I provided these changes within another issue. If it works for main views and doesn't work for some specific case - I think we could merge the PR and create a separate issue for the specific case to have an ability investigate this area more deeply.
WDYT?

Anyway I'm open for discussion - if there is the opinion that we can not merge the PR as is - then the solution should be  improved before merge.

**Comment by kittaakos** [comment]
> I tested the solution for `explorer`, `debug`, `search` and `scm` views only, so I didn't face the problem related to a custom view inside a custom view container. Please provide steps to reproduce or an example for testing.

Sure, I will. All cases should work after merging this PR.

> I provided these changes within another issue. If it works for main views and doesn't work for some specific case - I think we could merge the PR and create a separate issue for the specific case to have an ability investigate this area more deeply.
> WDYT?

No, I don't see it this way. Although nobody asked me for a review here, I spent my time investigating the VS Code source and proposed a reasonable code change. I tried your modification in-action; it did not work. Now I am creating a VS Code example for you. I do not want to come back for another follow-up PR review. This should be solved with this PR. Thank you for your understanding, and I am hoping to see this change in Theia ASAP. 👍

**Comment by kittaakos** [comment]
> Please provide steps to reproduce or an example for testing.

[Here](https://github.com/kittaakos/viewId) is the repo with the example, you can build one and try it.

In VS Code, you open the new view container, it has one view (it shows your user home) and you execute the `Do Something` command. It should show progress for two seconds:

screencast 2020-11-03 13-03-37

**Comment by RomanNikitenko** [comment]
> No, I don't see it this way.

OK, As I mentioned above:
> Anyway I'm open for discussion - if there is the opinion that we can not merge the PR as is - then the solution should be improved before merge.

> Although nobody asked me for a review here

But it's not true :-)
review

> I spent my time investigating the VS Code source and proposed a reasonable code change. 

>  Now I am creating a VS Code example for you.

thanks a lot!

**Comment by kittaakos** [comment]
> But it's not true :-)

I did not see the review request, sorry 😊. I just wanted to try the new feature.

**Comment by RomanNikitenko** [comment]
@kittaakos 
I tried your example - cloned and built it, then copied to the `theia/plugins` folder.
Unfortunately I got the error related to an activation the extension.

viewId
 
Could you take a look what's wrong in my steps?

**Comment by kittaakos** [comment]
> Could you take a look what's wrong in my steps?

I made some changes and verified it, it should work now. Thanks, @RomanNikitenko 🙏

**Comment by RomanNikitenko** [comment]
@kittaakos 
could you review when you have a chance

progress_test_view

**Review by kittaakos — APPROVED** [review]
I have verified with a custom view, it works. Thank you, @RomanNikitenko 👍

**Pull request merged**
